ANN2 fixes can add return types that are shadowed

### Summary

The fixes for [`missing-return-type-undocumented-public-function` (ANN201)](https://docs.astral.sh/ruff/rules/missing-return-type-undocumented-public-function/), [`missing-return-type-private-function` (ANN202)](https://docs.astral.sh/ruff/rules/missing-return-type-private-function/), [`missing-return-type-special-method` (ANN204)](https://docs.astral.sh/ruff/rules/missing-return-type-special-method/), [`missing-return-type-static-method` (ANN205)](https://docs.astral.sh/ruff/rules/missing-return-type-static-method/), and [`missing-return-type-class-method` (ANN206)](https://docs.astral.sh/ruff/rules/missing-return-type-class-method/) add return types without checking for shadowed variables, which can cause type-checking errors. They should use [`get_or_import_builtin_symbol`](https://github.com/astral-sh/ruff/blob/0.13.2/crates/ruff_linter/src/importer/mod.rs#L285) or similar. [Example](https://play.ruff.rs/816d2f85-9113-4039-8009-35420b66fe07):
```console
$ cat >ann2.py <<'# EOF'
from collections import UserString as str
from typing import override
def foo(): return "!"
def _foo(): return "!"
class C:
    @override
    def __str__(self): return "!"
    @staticmethod
    def foo(): return "!"
    @classmethod
    def bar(cls): return "!"
# EOF

$ ruff --isolated check --select ANN201,ANN202,ANN204,ANN205,ANN206 ann2.py --unsafe-fixes --fix
Found 5 errors (5 fixed, 0 remaining).

$ mypy ann2.py
ann2.py:3: error: Incompatible return value type (got "str", expected "UserString")  [return-value]
ann2.py:4: error: Incompatible return value type (got "str", expected "UserString")  [return-value]
ann2.py:7: error: Return type "UserString" of "__str__" incompatible with return type "str" in supertype "builtins.object"  [override]
ann2.py:7: error: Incompatible return value type (got "str", expected "UserString")  [return-value]
ann2.py:9: error: Incompatible return value type (got "str", expected "UserString")  [return-value]
ann2.py:11: error: Incompatible return value type (got "str", expected "UserString")  [return-value]
Found 6 errors in 1 file (checked 1 source file)
```

### Version

ruff 0.13.2 (b0bdf0334 2025-09-25)

--- a/crates/ruff_linter/resources/test/fixtures/flake8_annotations/shadowed_builtins.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_annotations/shadowed_builtins.py
@@ -1,0 +1,21 @@
+from collections import UserString as str
+from typing import override
+
+def foo():
+    return "!"
+
+def _foo():
+    return "!"
+
+class C:
+    @override
+    def __str__(self):
+        return "!"
+    
+    @staticmethod
+    def foo():
+        return "!"
+    
+    @classmethod
+    def bar(cls):
+        return "!"

--- a/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
+++ b/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
@@ -142,10 +142,7 @@ impl AutoPythonType {
                 });
                 Some((expr, vec![no_return_edit]))
             }
-            AutoPythonType::Atom(python_type) => {
-                let (expr, edits) = type_expr(python_type, checker, at)?;
-                Some((expr, edits))
-            }
+            AutoPythonType::Atom(python_type) => type_expr(python_type, checker, at),
             AutoPythonType::Union(python_types) => {
                 if target_version >= PythonVersion::PY310 {
                     // Aggregate all the individual types (e.g., `int`, `float`).
@@ -214,38 +211,14 @@ impl AutoPythonType {
 ///
 /// If the [`Expr`] relies on importing any external symbols, those imports will be returned as
 /// additional edits.
-fn type_expr(
+pub(crate) fn type_expr(
     python_type: PythonType,
     checker: &Checker,
     at: TextSize,
 ) -> Option<(Expr, Vec<Edit>)> {
     match python_type {
-        PythonType::String => {
-            let (edit, binding) = checker
-                .importer()
-                .get_or_import_builtin_symbol("str", at, checker.semantic())
-                .ok()?;
-            let expr = Expr::Name(ast::ExprName {
-                id: binding.into(),
-                range: TextRange::default(),
-                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
-                ctx: ExprContext::Load,
-            });
-            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
-        }
-        PythonType::Bytes => {
-            let (edit, binding) = checker
-                .importer()
-                .get_or_import_builtin_symbol("bytes", at, checker.semantic())
-                .ok()?;
-            let expr = Expr::Name(ast::ExprName {
-                id: binding.into(),
-                range: TextRange::default(),
-                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
-                ctx: ExprContext::Load,
-            });
-            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
-        }
+        PythonType::String => create_name_expr("str", checker, at),
+        PythonType::Bytes => create_name_expr("bytes", checker, at),
         PythonType::Number(number) => {
             let symbol = match number {
                 NumberLike::Integer => "int",
@@ -253,25 +226,10 @@ fn type_expr(
                 NumberLike::Complex => "complex",
                 NumberLike::Bool => "bool",
             };
-            let (edit, binding) = checker
-                .importer()
-                .get_or_import_builtin_symbol(symbol, at, checker.semantic())
-                .ok()?;
-            let expr = Expr::Name(ast::ExprName {
-                id: binding.into(),
-                range: TextRange::default(),
-                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
-                ctx: ExprContext::Load,
-            });
-            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+            create_name_expr(symbol, checker, at)
         }
         PythonType::None => {
-            let expr = Expr::Name(ast::ExprName {
-                id: "None".into(),
-                range: TextRange::default(),
-                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
-                ctx: ExprContext::Load,
-            });
+            let expr = Expr::NoneLiteral(ast::ExprNoneLiteral::default());
             Some((expr, vec![]))
         }
         PythonType::Ellipsis => None,
@@ -281,4 +239,24 @@ fn type_expr(
         PythonType::Tuple => None,
         PythonType::Generator => None,
     }
+}
+
+/// Helper function to create a name expression with proper builtin symbol handling.
+fn create_name_expr(
+    symbol: impl Into<Name>,
+    checker: &Checker,
+    at: TextSize,
+) -> Option<(Expr, Vec<Edit>)> {
+    let symbol_name = symbol.into();
+    let (edit, binding) = checker
+        .importer()
+        .get_or_import_builtin_symbol(symbol_name.as_str(), at, checker.semantic())
+        .ok()?;
+    let expr = Expr::Name(ast::ExprName {
+        id: binding.into(),
+        range: TextRange::default(),
+        node_index: ruff_python_ast::AtomicNodeIndex::NONE,
+        ctx: ExprContext::Load,
+    });
+    Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
 }

--- a/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
+++ b/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
@@ -216,9 +216,23 @@ pub(crate) fn type_expr(
     checker: &Checker,
     at: TextSize,
 ) -> Option<(Expr, Vec<Edit>)> {
+    fn name(name: &str, checker: &Checker, at: TextSize) -> Option<(Expr, Vec<Edit>)> {
+        let (edit, binding) = checker
+            .importer()
+            .get_or_import_builtin_symbol(name, at, checker.semantic())
+            .ok()?;
+        let expr = Expr::Name(ast::ExprName {
+            id: binding.into(),
+            range: TextRange::default(),
+            node_index: ruff_python_ast::AtomicNodeIndex::NONE,
+            ctx: ExprContext::Load,
+        });
+        Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+    }
+
     match python_type {
-        PythonType::String => create_name_expr("str", checker, at),
-        PythonType::Bytes => create_name_expr("bytes", checker, at),
+        PythonType::String => name("str", checker, at),
+        PythonType::Bytes => name("bytes", checker, at),
         PythonType::Number(number) => {
             let symbol = match number {
                 NumberLike::Integer => "int",
@@ -226,7 +240,7 @@ pub(crate) fn type_expr(
                 NumberLike::Complex => "complex",
                 NumberLike::Bool => "bool",
             };
-            create_name_expr(symbol, checker, at)
+            name(symbol, checker, at)
         }
         PythonType::None => {
             let expr = Expr::NoneLiteral(ast::ExprNoneLiteral::default());
@@ -239,24 +253,4 @@ pub(crate) fn type_expr(
         PythonType::Tuple => None,
         PythonType::Generator => None,
     }
-}
-
-/// Helper function to create a name expression with proper builtin symbol handling.
-fn create_name_expr(
-    symbol: impl Into<Name>,
-    checker: &Checker,
-    at: TextSize,
-) -> Option<(Expr, Vec<Edit>)> {
-    let symbol_name = symbol.into();
-    let (edit, binding) = checker
-        .importer()
-        .get_or_import_builtin_symbol(symbol_name.as_str(), at, checker.semantic())
-        .ok()?;
-    let expr = Expr::Name(ast::ExprName {
-        id: binding.into(),
-        range: TextRange::default(),
-        node_index: ruff_python_ast::AtomicNodeIndex::NONE,
-        ctx: ExprContext::Load,
-    });
-    Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
 }

--- a/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
+++ b/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
@@ -266,17 +266,13 @@ fn type_expr(
             Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
         }
         PythonType::None => {
-            let (edit, binding) = checker
-                .importer()
-                .get_or_import_builtin_symbol("None", at, checker.semantic())
-                .ok()?;
             let expr = Expr::Name(ast::ExprName {
-                id: binding.into(),
+                id: "None".into(),
                 range: TextRange::default(),
                 node_index: ruff_python_ast::AtomicNodeIndex::NONE,
                 ctx: ExprContext::Load,
             });
-            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+            Some((expr, vec![]))
         }
         PythonType::Ellipsis => None,
         PythonType::Dict => None,

--- a/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
+++ b/crates/ruff_linter/src/rules/flake8_annotations/helpers.rs
@@ -143,22 +143,27 @@ impl AutoPythonType {
                 Some((expr, vec![no_return_edit]))
             }
             AutoPythonType::Atom(python_type) => {
-                let expr = type_expr(python_type)?;
-                Some((expr, vec![]))
+                let (expr, edits) = type_expr(python_type, checker, at)?;
+                Some((expr, edits))
             }
             AutoPythonType::Union(python_types) => {
                 if target_version >= PythonVersion::PY310 {
                     // Aggregate all the individual types (e.g., `int`, `float`).
+                    let mut all_edits = Vec::new();
                     let names = python_types
                         .iter()
                         .sorted_unstable()
-                        .map(|python_type| type_expr(*python_type))
+                        .map(|python_type| {
+                            let (expr, mut edits) = type_expr(*python_type, checker, at)?;
+                            all_edits.append(&mut edits);
+                            Some(expr)
+                        })
                         .collect::<Option<Vec<_>>>()?;
 
                     // Wrap in a bitwise union (e.g., `int | float`).
                     let expr = pep_604_union(&names);
 
-                    Some((expr, vec![]))
+                    Some((expr, all_edits))
                 } else {
                     let python_types = python_types
                         .into_iter()
@@ -167,7 +172,7 @@ impl AutoPythonType {
 
                     match python_types.as_slice() {
                         [python_type, PythonType::None] | [PythonType::None, python_type] => {
-                            let element = type_expr(*python_type)?;
+                            let (element, mut edits) = type_expr(*python_type, checker, at)?;
 
                             // Ex) `Optional[int]`
                             let (optional_edit, binding) = checker
@@ -175,12 +180,18 @@ impl AutoPythonType {
                                 .import(at)
                                 .ok()?;
                             let expr = typing_optional(element, Name::from(binding));
-                            Some((expr, vec![optional_edit]))
+                            edits.push(optional_edit);
+                            Some((expr, edits))
                         }
                         _ => {
+                            let mut all_edits = Vec::new();
                             let elements = python_types
                                 .into_iter()
-                                .map(type_expr)
+                                .map(|python_type| {
+                                    let (expr, mut edits) = type_expr(python_type, checker, at)?;
+                                    all_edits.append(&mut edits);
+                                    Some(expr)
+                                })
                                 .collect::<Option<Vec<_>>>()?;
 
                             // Ex) `Union[int, str]`
@@ -189,7 +200,8 @@ impl AutoPythonType {
                                 .import(at)
                                 .ok()?;
                             let expr = typing_union(&elements, Name::from(binding));
-                            Some((expr, vec![union_edit]))
+                            all_edits.push(union_edit);
+                            Some((expr, all_edits))
                         }
                     }
                 }
@@ -199,26 +211,73 @@ impl AutoPythonType {
 }
 
 /// Given a [`PythonType`], return an [`Expr`] that resolves to that type.
-fn type_expr(python_type: PythonType) -> Option<Expr> {
-    fn name(name: &str) -> Expr {
-        Expr::Name(ast::ExprName {
-            id: name.into(),
-            range: TextRange::default(),
-            node_index: ruff_python_ast::AtomicNodeIndex::NONE,
-            ctx: ExprContext::Load,
-        })
-    }
-
+///
+/// If the [`Expr`] relies on importing any external symbols, those imports will be returned as
+/// additional edits.
+fn type_expr(
+    python_type: PythonType,
+    checker: &Checker,
+    at: TextSize,
+) -> Option<(Expr, Vec<Edit>)> {
     match python_type {
-        PythonType::String => Some(name("str")),
-        PythonType::Bytes => Some(name("bytes")),
-        PythonType::Number(number) => match number {
-            NumberLike::Integer => Some(name("int")),
-            NumberLike::Float => Some(name("float")),
-            NumberLike::Complex => Some(name("complex")),
-            NumberLike::Bool => Some(name("bool")),
-        },
-        PythonType::None => Some(name("None")),
+        PythonType::String => {
+            let (edit, binding) = checker
+                .importer()
+                .get_or_import_builtin_symbol("str", at, checker.semantic())
+                .ok()?;
+            let expr = Expr::Name(ast::ExprName {
+                id: binding.into(),
+                range: TextRange::default(),
+                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
+                ctx: ExprContext::Load,
+            });
+            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+        }
+        PythonType::Bytes => {
+            let (edit, binding) = checker
+                .importer()
+                .get_or_import_builtin_symbol("bytes", at, checker.semantic())
+                .ok()?;
+            let expr = Expr::Name(ast::ExprName {
+                id: binding.into(),
+                range: TextRange::default(),
+                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
+                ctx: ExprContext::Load,
+            });
+            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+        }
+        PythonType::Number(number) => {
+            let symbol = match number {
+                NumberLike::Integer => "int",
+                NumberLike::Float => "float",
+                NumberLike::Complex => "complex",
+                NumberLike::Bool => "bool",
+            };
+            let (edit, binding) = checker
+                .importer()
+                .get_or_import_builtin_symbol(symbol, at, checker.semantic())
+                .ok()?;
+            let expr = Expr::Name(ast::ExprName {
+                id: binding.into(),
+                range: TextRange::default(),
+                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
+                ctx: ExprContext::Load,
+            });
+            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+        }
+        PythonType::None => {
+            let (edit, binding) = checker
+                .importer()
+                .get_or_import_builtin_symbol("None", at, checker.semantic())
+                .ok()?;
+            let expr = Expr::Name(ast::ExprName {
+                id: binding.into(),
+                range: TextRange::default(),
+                node_index: ruff_python_ast::AtomicNodeIndex::NONE,
+                ctx: ExprContext::Load,
+            });
+            Some((expr, edit.map_or_else(Vec::new, |edit| vec![edit])))
+        }
         PythonType::Ellipsis => None,
         PythonType::Dict => None,
         PythonType::List => None,

--- a/crates/ruff_linter/src/rules/flake8_annotations/mod.rs
+++ b/crates/ruff_linter/src/rules/flake8_annotations/mod.rs
@@ -229,4 +229,20 @@ mod tests {
         assert_diagnostics!(diagnostics);
         Ok(())
     }
+
+    #[test]
+    fn shadowed_builtins() -> Result<()> {
+        let diagnostics = test_path(
+            Path::new("flake8_annotations/shadowed_builtins.py"),
+            &LinterSettings::for_rules(vec![
+                Rule::MissingReturnTypeUndocumentedPublicFunction,
+                Rule::MissingReturnTypePrivateFunction,
+                Rule::MissingReturnTypeSpecialMethod,
+                Rule::MissingReturnTypeStaticMethod,
+                Rule::MissingReturnTypeClassMethod,
+            ]),
+        )?;
+        assert_diagnostics!(diagnostics);
+        Ok(())
+    }
 }

--- a/crates/ruff_linter/src/rules/flake8_annotations/rules/definition.rs
+++ b/crates/ruff_linter/src/rules/flake8_annotations/rules/definition.rs
@@ -4,13 +4,14 @@ use ruff_python_ast::identifier::Identifier;
 use ruff_python_ast::visitor::Visitor;
 use ruff_python_ast::{self as ast, Expr, Stmt};
 use ruff_python_semantic::Definition;
+use ruff_python_semantic::analyze::type_inference::{NumberLike, PythonType};
 use ruff_python_semantic::analyze::visibility;
 use ruff_python_stdlib::typing::simple_magic_return_type;
 use ruff_text_size::Ranged;
 
 use crate::checkers::ast::{Checker, DiagnosticGuard};
 use crate::registry::Rule;
-use crate::rules::flake8_annotations::helpers::auto_return_type;
+use crate::rules::flake8_annotations::helpers::{auto_return_type, type_expr};
 use crate::rules::ruff::typing::type_hint_resolves_to_any;
 use crate::{Edit, Fix, FixAvailability, Violation};
 
@@ -825,11 +826,31 @@ pub(crate) fn definition(
                     },
                     function.identifier(),
                 );
-                if let Some(return_type) = return_type {
-                    diagnostic.set_fix(Fix::unsafe_edit(Edit::insertion(
-                        format!(" -> {return_type}"),
-                        function.parameters.end(),
-                    )));
+                if let Some(return_type_str) = return_type {
+                    // Convert the simple return type to a proper expression that handles shadowed builtins
+                    let python_type = match return_type_str {
+                        "str" => PythonType::String,
+                        "bytes" => PythonType::Bytes,
+                        "int" => PythonType::Number(NumberLike::Integer),
+                        "float" => PythonType::Number(NumberLike::Float),
+                        "complex" => PythonType::Number(NumberLike::Complex),
+                        "bool" => PythonType::Number(NumberLike::Bool),
+                        "None" => PythonType::None,
+                        _ => return, // Unknown type, skip
+                    };
+
+                    if let Some((expr, edits)) =
+                        type_expr(python_type, checker, function.parameters.start())
+                    {
+                        let return_type_expr = checker.generator().expr(&expr);
+                        diagnostic.set_fix(Fix::unsafe_edits(
+                            Edit::insertion(
+                                format!(" -> {return_type_expr}"),
+                                function.parameters.end(),
+                            ),
+                            edits,
+                        ));
+                    }
                 }
                 diagnostics.push(diagnostic);
             }

